app/src/main/java/com/keylesspalace/tusky/BaseActivity.java

```diff
@@ -177,6 +192,36 @@ protected void setPullNotificationCheckInterval(long minutes) {
                 .scheduleAsync();
     }
 
+    protected void downloadFile(String url) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
+                ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                        != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this,
+                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
+                    PERMISSIONS_REQUEST_WRITE_EXTERNAL_STORAGE);
+        } else {
+            String filename = new File(url).getName();
+
+            String toastText = String.format(getResources().getString(R.string.download_image), filename);
+            Toast.makeText(getApplicationContext(), toastText, Toast.LENGTH_SHORT).show();
+
+            DownloadManager downloadManager = (DownloadManager) getSystemService(Context.DOWNLOAD_SERVICE);
+            DownloadManager.Request request = new DownloadManager.Request(Uri.parse(url));
+            request.allowScanningByMediaScanner();
+            request.setDestinationInExternalPublicDir(Environment.DIRECTORY_PICTURES,
+                    getString(R.string.app_name) + "/" + filename);
```

Comment:
Making file paths by hand is usually a bad idea but maybe that's okay?

Reply:
This is just moved from `ViewVideoActivity` as well. :smile: